### ISSUE TITLE
@Html.Editor("") does not support nested properties correctly

### ISSUE THREAD

**Issue by Inspyro** [aspnet/Mvc#6639]
Hi, 
We are using tag helpers in MVC Core so for us it was logical to call `Html.Editor("Model.SubModel.Property")` instead of using a lambda expression.

Now we found out that using this syntax does not really support nested properties, as it always creates textboxes (and does not consider the templates / datatypes).

**Minimal repro sample**
[AspCoreRazorEditorBug.zip](https://github.com/aspnet/Mvc/files/1211770/AspCoreRazorEditorBug.zip)

View
```
@Html.EditorFor(m => m.MyNestedModel.PackagingSize)
@Html.Editor("MyNestedModel.PackagingSize")
```

* `Html.EditorFor` works as expected and renders a select control
* `Html.Editor` renderes a textbox for the enum property (with the correct label)

EditorTemplate
```
@using System
@model Enum
<select class="form-control" asp-for="@Model" asp-items="@Html.GetEnumSelectList(ViewData.ModelMetadata.UnderlyingOrModelType)">
</select>
```

Nested Model
```
public class MyNestedModel
{
    [EnumDataType(typeof(PackingSize))]
    public PackingSize PackagingSize;
}
```

**Comment by Inspyro** [comment]
I found out that we could call 
`_htmlHelper.GetType().GetMethod("GenerateEditor", BindingFlags.Instance | BindingFlags.NonPublic);
      var result = (IHtmlContent) generateEditorMethod.Invoke(
          _htmlHelper,
          new[] { modelExplorer, htmlFieldName, templateName, additionalViewData });`

the "GenerateEditor" method with the For.ModelExplorer instead of the For.Name (this is what Html.Editor does internally, but apparently wrong)
this would be a workaround, but I would still consider it a bug in mvc core

**Comment by dougbu** [comment]
Please follow up at #4821. This is a duplicate of that issue.

#### Background
This issue is one symptom of a longstanding restriction in MVC, going back at least to MVC 5.2.3. The behaviour is still visible in 2.0.0-preview2-final and the latest dev code.

As you've seen, expression names can only reach down one level in a model. `@Html.Editor("Property")` works fine but `@Html.Editor("Property.InnerProperty")` does not (especially when `Model` is `null`). It's possible to improve this but the bug (#4821) has been on our backlog for a while.

Workarounds all involve either using lambda-based HTML helpers or limiting expression names to a simple property name. For example, you could create a template for the `MyNestedModel` type instead.

**Comment by dougbu** [comment]
FYI an additional workaround is available in the upcoming `2.0.0` release: The view correctly finds the (uninitialized) `PackagingSize` property if the action is changed to pass a non-`null` model:
``` c#
var model = new Container
{
    MyClass = new MyClass(),
};

return View(model);
```

**Comment by fschmied** [comment]
@dougbu We encountered the problem when trying to invoke the `Editor` HTML helper from within a tag helper. The `ModelExpression` used by tag helpers only provides the expression as a string, not as a Lambda, which is why we couldn't use the `EditorFor` overload. I also think the non-`null` model workaround won't work for the tag helper issue, or will it?

I've created a dedicated issue for this problem: #6645.
